File: src/main/java/de/md5lukas/commands/BaseCommand.java
package de.md5lukas.commands;

import com.google.common.base.Preconditions;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Class that is the root of all sub-commands and gets registered as the command executor and tab completer
 */
public abstract class BaseCommand extends SubCommand implements CommandExecutor, TabCompleter {

    /**
     * The name of this command, the same as in the plugin.yml. The name in the plugin.yml has to be lowercase
     *
     * @param name The name of this command
     */
    public BaseCommand(String name) {
        super(name);
    }

    @Override
    public final boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        this.onCommand(sender, new LinkedList<>(Arrays.asList(args)));
        return true;
    }

    @Override
    public final List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        return this.onTabComplete(sender, new LinkedList<>(Arrays.asList(args)));
    }

    /**
     * Registers this command as the executor and tab completer for the given command name
     *
     * @param plugin The plugin to use to register this command
     */
    @SuppressWarnings("ConstantConditions")
    public final void register(JavaPlugin plugin) {
        Preconditions
                .checkNotNull(plugin.getCommand(getName()), "It seems like the command with the name %s has not been registered in the plugin.yml", getName());
        plugin.getCommand(getName()).setExecutor(this);
        plugin.getCommand(getName()).setTabCompleter(this);
    }
}

File: src/main/kotlin/de/md5lukas/commands/BaseCommand.kt
package de.md5lukas.commands

import de.md5lukas.commands.dsl.options.BaseCommandOptions
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

class BaseCommand(options: BaseCommandOptions) : SubCommand(options, options), CommandExecutor, TabCompleter {

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<out String>
    ): Boolean {
        val linkedArgs = LinkedList<String>()
        args.forEach { linkedArgs.add(it) }

        this.onCommand(sender, linkedArgs, "")

        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        alias: String,
        args: Array<out String>
    ): MutableList<String> {
        val linkedArgs = LinkedList<String>()
        args.forEach { linkedArgs.add(it) }

        return ArrayList(this.onTabComplete(sender, linkedArgs))
    }

    fun register(plugin: JavaPlugin) {
        val pc = plugin.getCommand(name)
            ?: throw IllegalStateException("The plugin ${plugin.name} has not a command registered with the name $name")
        pc.setExecutor(this)
        pc.tabCompleter = this
    }
}